topic-operator/design/topic-store.md

```diff
@@ -0,0 +1,122 @@
+# Topic store (new Kafka Streams based implementation) - design document
+
+Topic store represents a persistent data store where the operator can store its copy of the 
+topic state that won't be modified by either K8S or Kafka.
+Currently we have ZooKeeper based working implementation, but with ZooKeeper being removed 
+as part of KIP-500, we decided to implement the store directly in/on Kafka, its Streams
+extension to be exact.
+
+TopicStore interface is a simple async CRUD API.
+
+```
+interface TopicStore {
+
+    /**
+     * Asynchronously get the topic with the given name
+     * completing the returned future when done.
+     * If no topic with the given name exists, the future will complete with
+     * a null result.
+     * @param name The name of the topic.
+     * @return A future which completes with the given topic.
+     */
+    Future<Topic> read(TopicName name);
+
+    /**
+     * Asynchronously persist the given topic in the store
+     * completing the returned future when done.
+     * If a topic with the given name already exists, the future will complete with an
+     * {@link EntityExistsException}.
+     * @param topic The topic.
+     * @return A future which completes when the given topic has been created.
+     */
+    Future<Void> create(Topic topic);
+
+    /**
+     * Asynchronously update the given topic in the store
+     * completing the returned future when done.
+     * If no topic with the given name exists, the future will complete with a
+     * {@link NoSuchEntityExistsException}.
+     * @param topic The topic.
+     * @return A future which completes when the given topic has been updated.
+     */
+    Future<Void> update(Topic topic);
+
+    /**
+     * Asynchronously delete the given topic from the store
+     * completing the returned future when done.
+     * If no topic with the given name exists, the future will complete with a
+     * {@link NoSuchEntityExistsException}.
+     * @param topic The topic.
+     * @return A future which completes when the given topic has been deleted.
+     */
+    Future<Void> delete(TopicName topic);
+}
+```
+
+The TopicStore is used from TopicOperator, which is used from web/Vert.x invocations
+and ZooKeeper watcher callbacks. The later - ZooKeeper watcher callbacks - also need
```

Comment:
Slightly pedantic point: I don't think the `TopicStore` interface is used by the watchers, though they _do_ use the underlying ZK client abstraction.

Reply:
Re-reading this, I did mean TO being used from Vert.x and ZK watchers.
But I guess it should be more clear, so you don't think TS is directly used by the watchers.
Any suggestion how to reword / rephrase this?

Reply:
Something like
```
The TopicStore is only used by the TopicOperator. 
The ZooKeeper watcher callbacks trigger TopicOperator reconciliations, and this notification mechanism will also need to be replaced before KIP-500 is fully implemented.
```